elastic/kibana
# [Ingest Pipelines] Add unsaved changes prompt

Closes https://github.com/elastic/kibana/issues/183503

## Summary

Whenever a user inputs data into the edit/create ingest pipeline and navigates away, all the data the entered will be lost. This PR adds a prompt to confirm with them if they are sure they want to navigate away only when they have pending changes in the form.

**Note to reviewers:**
This PR introduces a new package with a hook that can be useful for preventing the user from navigating away from a form when having a dirty form:

```typescript
import { useUnsavedChangesPrompt } from '@kbn/unsaved-changes-prompt';

export const SampleForm = ({ openConfirm, history, http, navigateToUrl }) => {
  const { form } = useForm();
  const isFormDirty = useFormIsModified({ form });
  
    useUnsavedChangesPrompt({
      hasUnsavedChanges: isFormDirty,
      openConfirm,
      history,
      http,
      navigateToUrl,
    });

  return (
    <Form {...form}>
       ....
    </Form>
  );
};
```


https://github.com/elastic/kibana/assets/1191206/e32ace3f-bda6-4b12-b524-a95589927677



##### How to test

* Start up kibana and es
* Navigate to Stack management -> Ingest pipelines
* Create a new pipeline or edit an existing pipeline
* Verify that:
  * Making changes and navigating away shows a prompt
  * If no changes were made, the user can navigate away as they normally would
